### PR TITLE
Add keys as a child proxy resource under vaults (AKV's ARM contract).

### PR DESCRIPTION
**Pull request by adarce**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [ ] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by adarce** [comment]
@heaths , @schaabs , @fengzhou-msft : I see that the required "**automation - sdk (SDK Python)**" check fails during the Docker container initialization step. What team should this issue be escalated to?

**Comment by fengzhou-msft** [comment]
@NullMDR can you help check the automation - sdk failures?

**Comment by adarce** [comment]
Thanks for investigating.

@fengzhou-msft : Is this issue something that I can fix? If so, how do I do this?

**Comment by heaths** [comment]
Why are you modifying a GA'd swagger? Can you describe what this PR is for? Is something missing, or are you adding features, which should go into a new service version?

**Comment by adarce** [comment]
@heaths : I am modifying the most recent stable API version for KeyVault's control plane. This PR is additive (adding new resource types), so as per [this ARM documentation for API versions](https://microsoft.sharepoint.com/:w:/t/azureresourcemanagerteam/EWXsAQ1yx25KkyYCeeWGUgwBSBxEdUDEbHi6FZ__U8EOQw?e=xEcppo), these non-breaking API changes do not require a new API version.

Overall, these new APIs will allow customers to create keys via ARM templates. KeyVault's control plane will be proxying such requests to KeyVault's data plane. By exposing data plane APIs via ARM templates, we enable seamless UX workflows for partner teams such as Azure Disk Encryption.

**Comment by heaths** [comment]
@RandalliLama can you get someone from the service team to review as well?

File: specification/keyvault/resource-manager/readme.md
Comment (heaths) [inline specification/keyvault/resource-manager/readme.md:41]: Are you sure that all generators are pointing to this readme.md for configuration? At least for data plane, we point at the JSON files directly. I.e. this should be merged into `keyvault.json` to work with existing tooling and configuration.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (heaths) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:288]: Question: what's the use case for this one? ListKeys I can see being used in an ARM function to query for a specific key (must like getting a storage account key), but the versions themselves are, for all intent to the customer, opaque versions. Referring to a specific key (`Keys_GetVersion`) makes sense to refer to a fixed version, but why support listing them in ARM?

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (heaths) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:353]: There should already be an error type from keyvault.json. IF you don't merge this into keyvault.json (recommended), I recommend breaking out the error type into another file and use an external ref (see what I did in the data-plane for KV 7.1) to avoid unnecessarily functional duplication in generated code.

Comment (heaths) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:353]: Same for other shared definitions below.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (heaths) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:376]: Should these be imported from the data plane to make sure they stay in sync, or is that necessarily important here as long as the useful data from keys can be queried in ARM functions?

For example, KeyProperties in 7.1 also has `recoveryDays` and a couple more values for `recoveryLevel`, IIRC.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (heaths) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:439]: Not "oct" for Managed HSM?

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (heaths) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:451]: "backup" and "restore" not needed here?

**Review by heaths — APPROVED** [review]
Perhaps it's unnecessary, but we might consider sharing JWK and other related definitions with data-plane. Or perhaps the idea is not to return everything in a response for ARM functions? The defs are already out of sync with what shipped in 7.1 GA and in 7.2 preview.

**Comment by NullMDR** [comment]
@adarce I could help to update your base branch. Please let me know if you want.

**Comment by adarce** [comment]
@NullMDR : Yes, please assist.

File: specification/keyvault/resource-manager/readme.md
Comment (adarce) [reply specification/keyvault/resource-manager/readme.md:41]: Thanks for the feedback - this is my first time authoring a swagger spec. I'll look into reverting this and updating `keyvault.json` instead. (I wanted to avoid updating any existing files and only add new files, so as to avoid any breaking changes.)

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (adarce) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:288]: Good question. The ListKeys/ListKeyVersions APIs are not intended to be used in ARM templates (they are paginated by continuation tokens, which I don't think ARM templates support). Instead, they are intended to be used by UX in the Azure Portal.

The scenario is that a partner team (e.g.: Azure Disk Encryption) will allow customers to provide customer-managed keys by opening a picker control owned by the Key Vault team. This picker will have a dropdown or grid that will list all keys in a given vault, as well as list all key versions for a given key. This picker will call these APIs via ARM instead of directly accessing the data plane because the two code paths use different permission models:
- **Direct data plane APIs:** Permissions to list keys or key versions are configured either via legacy access policies, or via RBAC DataActions.
- **Data plane APIs proxied via control plane:** Permissions to list keys or key versions are configured via RBAC Actions, and can bypass firewall.

The reasoning behind using a separate permission model for these new APIs instead of overloading the existing RBAC DataActions:
- **Fundamentally different APIs:** Although both accomplish the same thing, CreateKey Action is a management operation (typically, a one-time only setup). The CreateKey DataAction is a data operation (used frequently).
- **Restricted elevation of privilege:** We want users with permissions to create vaults via ARM to implicitly also have permission to indirectly create keys via ARM. This happens because owners/contributors have permissions that match `Microsoft.KeyVault/vaults/*` RBAC Actions, which includes the CreateKey Action. However, we don't want these users to automatically be able to create keys directly via the data plane.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (adarce) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:439]: Key Vault doesn't currently support "oct", but perhaps it's a good idea to include this in the enum anyways in case we eventually do. I'll add it.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (adarce) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:451]: No, backup and restore aren't JsonWebKeyOperations (although they are key permissions that can be configured in a vault's access policy). I included an additional operation ("import") because it will be allowed for a different upcoming feature.

Compare with the swagger spec from data plane: https://github.com/Azure/azure-rest-api-specs/blob/6b46480ebd9b2f2c7220f2e9a9c6754dd524e2d1/specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.0/keyvault.json#L4782-L4800

Screenshot from Azure Portal:
image

**Comment by adarce** [comment]
> Perhaps it's unnecessary, but we might consider sharing JWK and other related definitions with data-plane. Or perhaps the idea is not to return everything in a response for ARM functions? The defs are already out of sync with what shipped in 7.1 GA and in 7.2 preview.

I'm willing to refactor things to avoid unnecessary duplication, but I want to avoid making any breaking changes to existing APIs. Also yes: for the ARM APIs, the keys returned from ARM will contain a subset of the properties exposed by keys directly returned from the data plane (e.g.: the full JWK won't be returned in the ARM APIs, so RSA keys won't have their modulus/exponent exposed for example).

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (heaths) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:451]: Thanks. I seemed to have confused key_ops for key permissions.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (majastrz) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:508]: >tags [](start = 9, length = 4)

Are keys going to be tracked or proxy resources in ARM?

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (majastrz) [inline specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:211]: >get [](start = 7, length = 3)

I don't see a PUT. Does this mean that a PUT on /keys/{name} will automatically create a /keys/{name}/version/{keyVersion} resource in the backend?

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (majastrz) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:508]: The reason I ask is because tags shouldn't really be used on proxy resources.

---
In reply to: [483360952](https://github.com/Azure/azure-rest-api-specs/pull/10643#discussion_r483360952) [](ancestors = 483360952)

Comment (adarce) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:508]: Yes, keys are going to be proxy resources in ARM, in the same way that secrets are child proxy resources under vaults in ARM. See here for comparison: https://github.com/Azure/azure-rest-api-specs/blob/9603e8c3886893f114df4a2dc844d175c3f2b1c6/specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/secrets.json#L361-L369


Note that Azure Key Vault's data plane supports its own tags for all objects that can be stored in a vault (keys, secrets, certificates, and managed storage account keys).

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (adarce) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:211]: Yes, that's correct.

File: specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json
Comment (majastrz) [reply specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json:508]: I see.

**Review by majastrz — APPROVED** [review]
Signed off from ARM side.

**Comment by fengzhou-msft** [comment]
@NullMDR please help update the base branch.

**Comment by fengzhou-msft** [comment]
@adarce the base branch has conflicts with the master branch. Do you need to keep the changes for the base branch or just reset it to master branch?

**Comment by adarce** [comment]
> @adarce the base branch has conflicts with the master branch. Do you need to keep the changes for the base branch or just reset it to master branch?

Whichever option is easier. I'm not familiar with the differences between the base branch and the master branch. My goal is simply to make additive, non-breaking changes to the existing 2019-09-01 API version. Given that the ARM team already signed off on this PR, I'd like to preserve this PR and not abandon it. Is it possible to update this PR in-place, without creating a new one?

**Comment by PhoenixHe-msft** [comment]
@heaths would you please help to resolve the conflict between https://github.com/Azure/azure-rest-api-specs/tree/dev-keyvault-Microsoft.KeyVault-2019-09-01 and master?

**Comment by heaths** [comment]
Someone from the service team should take a look. /cc @RandalliLama

**Comment by RandalliLama** [comment]
@heaths @adarce is from the service team.  :). If the two of you are in sync then I approve.

**Comment by heaths** [comment]
I mean someone other than the person who implemented the change - i.e. someone that knows the service side better and can verify the swagger matches the implementation (like any normal code review). I can only watch for issues that would affect the SDK (or any generated client, for that matter).

**Review by AshokCMSFT — APPROVED** [review]

**Review by AshokCMSFT — APPROVED** [review]

**Comment by AshokCMSFT** [comment]
@heaths I have reviewed Adriano's changes and approved it

**Comment by mikhailshilkov** [comment]
Is there no DELETE operation available? Any plans to add it?

**Comment by adarce** [comment]
> Is there no DELETE operation available? Any plans to add it?

I'm not aware of any specific plans to support the key deletion scenario in particular, but generally speaking we plan to support more data plane operations in ARM templates.

Feel free to leave suggestions on our user voice forum, which helps us to triage/prioritize customer asks for future feature work: https://feedback.azure.com/forums/906355-azure-key-vault

**Comment by adarce** [comment]
@heaths, @jlichwa : I have rebased my PR onto the latest from master, and all required checks are now passing. Is there anything else I need to do before you can merge the PR?

**Review by jlichwa — APPROVED** [review]

**Comment by heaths** [comment]
@fengzhou-msft is the one who will have to merge the PR.

**Comment by adarce** [comment]
@fengzhou-msft : I have rebased my PR onto the latest from master, and all required checks are now passing. Is there anything else I need to do before you can merge the PR?

**Pull request merged**
